src/util/bip152.rs

```diff
@@ -0,0 +1,438 @@
+// Rust Bitcoin Library
+//
+// To the extent possible under law, the author(s) have dedicated all
+// copyright and related and neighboring rights to this software to
+// the public domain worldwide. This software is distributed without
+// any warranty.
+//
+// You should have received a copy of the CC0 Public Domain Dedication
+// along with this software.
+// If not, see <http://creativecommons.org/publicdomain/zero/1.0/>.
+//
+
+//! BIP152 Compact Blocks
+//!
+//! Implementation of compact blocks data structure and algorithms.
+//!
+
+use std::{convert, error, fmt, io, mem};
+
+use hashes::{hex, sha256, siphash24, Hash};
+
+use hash_types::BlockHash;
+use blockdata::block::{Block, BlockHeader};
+use blockdata::transaction::Transaction;
+use consensus::encode::{self, Decodable, Encodable, VarInt};
+use util::endian;
+
+/// A BIP-152 error
+#[derive(Clone, PartialEq, Eq, Debug)]
+pub enum Error {
+    /// An unknown version number was used.
+    UnknownVersion,
+    /// A transaction index is requested that is out
+    /// of range from the corresponding block.
+    TxIndexOutOfRange(u64),
+    /// The prefill slice provided was invalid.
+    InvalidPrefill,
+}
+
+impl fmt::Display for Error {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match *self {
+            Error::UnknownVersion => write!(f, "an unknown version number was used"),
+            Error::TxIndexOutOfRange(i) => write!(f, "a transaction index is requested that is \
+                out of range from the corresponding block: {}", i,
+            ),
+            Error::InvalidPrefill => write!(f, "the prefill slice provided was invalid"),
+        }
+    }
+}
+
+impl error::Error for Error {
+    fn cause(&self) -> Option<&error::Error> {
+        None
+    }
+
+    fn description(&self) -> &str {
+        "description() is deprecated; use Display"
+    }
+}
+
+/// A [PrefilledTransaction] structure is used in [HeaderAndShortIDs] to
+/// provide a list of a few transactions explicitly.
+#[derive(PartialEq, Eq, Clone, Debug)]
+pub struct PrefilledTransaction {
+    /// The index of the transaction in the block.
+    ///
+    /// This field is differentially encoded relative to the previous
+    /// prefilled transaction as described as follows:
+    ///
+    /// > Several uses of CompactSize below are "differentially encoded". For
+    /// > these, instead of using raw indexes, the number encoded is the
+    /// > difference between the current index and the previous index, minus one.
+    /// > For example, a first index of 0 implies a real index of 0, a second
+    /// > index of 0 thereafter refers to a real index of 1, etc.
+    pub idx: u64,
```

Comment:
Nit: this could be a u32 or maybe even a u16 (won't have 65535 transactions per block anytime soon)

Reply:
Agreed, especially since [Core uses a `uint16_t` too](https://github.com/bitcoin/bitcoin/blob/6a726cb534ed471010bf3f466732874b2a1b22e5/src/blockencodings.h#L71).

Reply:
fixed